### PR TITLE
Yml manifest implementation

### PR DESCRIPTION
**Pull request by kris-nova**
Pull Request for work defined in [PR 432](https://github.com/kubernetes/kops/pull/432)

Hey @justinsb, can we maybe try to get this in before the release later this week?
@chrislovecnm - Can you take a look and make sure this meets what you are thinking?
@timothysc - See the viper implementation ;)

Cheers

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -13,6 +13,9 @@ endif
 kops: gobindata
 	GO15VENDOREXPERIMENT=1 go install -ldflags "-X main.BuildVersion=${VERSION}" k8s.io/kops/cmd/kops/...
 
+cobra-test:
+	go test k8s.io/kops/cmd/kops
```
Comment (chrislovecnm) [inline Makefile]: Do we want to roll this up to a full test target?

File: cmd/kops/create_cluster.go [hunk 2]
```diff
@@ -86,6 +87,8 @@ func init() {
 	cmd.Flags().StringVar(&createCluster.AdminAccess, "admin-access", "", "Restrict access to admin endpoints (SSH, HTTPS) to this CIDR.  If not set, access will not be restricted by IP.")
 
 	cmd.Flags().BoolVar(&createCluster.AssociatePublicIP, "associate-public-ip", true, "Specify --associate-public-ip=[true|false] to enable/disable association of public IP for master ASG and nodes. Default is 'true'.")
+
+	cmd.Flags().StringVarP(&createCluster.Manifest, "manifest", "f", "", "Optional .yml cluster manifest")
```
Comment (chrislovecnm) [inline cmd/kops/create_cluster.go]: I would steal the verbiage from `kubectl`: -f, 

We are not going to do a directory at this point. 
```
--filename=[]: Filename, directory, or URL to a file identifying the resource to get from a server.
```

File: cmd/kops/create_cluster.go [hunk 3]
```diff
@@ -94,6 +97,21 @@ func (c *CreateClusterCmd) Run(args []string) error {
 		return err
 	}
 
+	if c.Manifest != "" {
+		clusterConfig, err := getManifest(c.Manifest)
+		if err != nil {
+			return err
```
Comment (chrislovecnm) [inline cmd/kops/create_cluster.go]: Should we log a message here?

File: cmd/kops/create_cluster.go [hunk 4]
```diff
@@ -94,6 +97,21 @@ func (c *CreateClusterCmd) Run(args []string) error {
...
+		}
+		// This is a WeakDecode() that won't throw an error if a directive does not exist.
+		clusterConfig.Unmarshal(c)
+		c.Zones = clusterConfig.GetString("Zones")
+		if c.Zones == "" {
```
Comment (chrislovecnm) [inline cmd/kops/create_cluster.go]: When would `c.Zones == ""`???

File: cmd/kops/delete_cluster.go [hunk 5]
```diff
@@ -42,6 +43,8 @@ func init() {
 	cmd.Flags().BoolVar(&deleteCluster.External, "external", false, "Delete an external cluster")
 
 	cmd.Flags().StringVar(&deleteCluster.Region, "region", "", "region")
+
+	cmd.Flags().StringVarP(&deleteCluster.Manifest, "manifest", "f", "", "Optional .yml cluster manifest")
```
Comment (chrislovecnm) [inline cmd/kops/delete_cluster.go]: Should we store the command messages in one file? I am a huge fan of duplicated strings ... not.  This is probably another PR ...

File: cmd/kops/manifest_test.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+// to enable these tests remove the `t.SkipNow()` methods, and flesh out manifest_example.yml with
+// a valid test environment configuration.
+// - or -
+// Create a new manifest testing .yml file and update the manifest var
+
+import (
+	"testing"
+)
+
+var manifest = "../../manifest_example.yml"
```
Comment (chrislovecnm) [inline cmd/kops/manifest_test.go]: Should we allow this to be an ENV VAR?  That way we can dynamically set our test files?

File: cmd/kops/create_cluster.go (hunk 2 above)
Comment (kris-nova) [inline cmd/kops/create_cluster.go]: My thoughts..

All CRUD operations should use :

--filename=[]: Filename, or directory of a file identifying a .yml manifest.

File: Makefile (hunk 1 above)
Comment (kris-nova) [inline Makefile]: There already is a test: target in the Makefile. I was thinking we can combine them later if needed, but that is probably outside the scope of this PR, and requires a broader conversation on testing.

For now - this target just runs any go tests in the cmd/kops directory.

File: cmd/kops/create_cluster.go (hunk 3 above)
Comment (kris-nova) [inline cmd/kops/create_cluster.go]: Nope, lets return the error and let something further upstream handle reporting the error.

File: cmd/kops/create_cluster.go (hunk 4 above)
Comment (kris-nova) [inline cmd/kops/create_cluster.go]: The zones directive currently supports **YAML strings** and **YAML lists**.. If a list is defined the GetString() function will return an empty string, we use that to then look for a YAML list with GetStringSlice()

File: cmd/kops/delete_cluster.go (hunk 5 above)
Comment (kris-nova) [inline cmd/kops/delete_cluster.go]: I think the way cobra separates concerns is going to force us defining this for every command. Yes it's duplicate code, but it's probably a good thing as each implementation is subject to change.

File: cmd/kops/manifest_test.go (hunk 6 above)
Comment (kris-nova) [inline cmd/kops/manifest_test.go]: Yes. But I think we should have a default so the user isn't forced to set it. I can add this to the next commit.

File: cmd/kops/create_cluster.go (hunk 2 above)
Comment (chrislovecnm) [inline cmd/kops/create_cluster.go]: just remove directory.  And double check the different commands in kubectl.

File: cmd/kops/create_cluster.go (hunk 3 above)
Comment (chrislovecnm) [inline cmd/kops/create_cluster.go]: Yah you mentioned Cobra handles it.  GI Joe joke in there somewhere.

File: cmd/kops/create_cluster.go [hunk 7]
```diff
@@ -88,7 +88,7 @@ func init() {
 
 	cmd.Flags().BoolVar(&createCluster.AssociatePublicIP, "associate-public-ip", true, "Specify --associate-public-ip=[true|false] to enable/disable association of public IP for master ASG and nodes. Default is 'true'.")
 
-	cmd.Flags().StringVarP(&createCluster.Manifest, "manifest", "f", "", "Optional .yml cluster manifest")
+	cmd.Flags().StringVarP(&createCluster.Manifest, "filename", "f", "", "Filename, or directory of a file identifying a .yml manifest.")
```
Comment (chrislovecnm) [inline cmd/kops/create_cluster.go]: `- directory` - that was my typo

File: cmd/kops/update_cluster.go [hunk 8]
```diff
@@ -45,7 +45,7 @@ func init() {
 	cmd.Flags().StringVar(&updateCluster.SSHPublicKey, "ssh-public-key", "", "SSH public key to use (deprecated: use kops create secret instead)")
 	cmd.Flags().StringVar(&updateCluster.OutDir, "out", "", "Path to write any local output")
 
-	cmd.Flags().StringVarP(&updateCluster.Manifest, "manifest", "f", "", "Optional .yml cluster manifest")
+	cmd.Flags().StringVarP(&updateCluster.Manifest, "filename", "f", "", "Filename, or directory of a file identifying a .yml manifest.")
```
Comment (chrislovecnm) [inline cmd/kops/update_cluster.go]: Same on this one ... remove the work "directory"

File: cmd/kops/create_cluster.go (hunk 2 above)
Comment (kris-nova) [inline cmd/kops/create_cluster.go]: @chrislovecnm - "directory" has been removed.. These look okay now?

File: cmd/kops/create_cluster.go (hunk 3 above)
Comment (chrislovecnm) [inline cmd/kops/create_cluster.go]: Yay

**Comment by kris-nova** [comment]
Nope @chrislovecnm - we are all good on my side

**Comment by chrislovecnm** [comment]
@justinsb - status on this?

**Comment by kris-nova** [comment]
Closing - this PR is moot at this point. 

Cheers gentlemen.

**Pull request closed without merging**